app/src/main/java/org/wikipedia/random/RandomViewModel.kt

```diff
@@ -0,0 +1,103 @@
+package org.wikipedia.random
+
+import android.os.Bundle
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.ViewModelProvider
+import androidx.lifecycle.viewModelScope
+import kotlinx.coroutines.*
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import org.wikipedia.Constants
+import org.wikipedia.WikipediaApp
+import org.wikipedia.analytics.RandomizerFunnel
+import org.wikipedia.database.AppDatabase
+import org.wikipedia.dataclient.WikiSite
+import org.wikipedia.page.PageTitle
+import org.wikipedia.util.log.L
+
+class RandomViewModel(bundle: Bundle) : ViewModel() {
+
+	private val funnel: RandomizerFunnel
+	private var saveShareJob: Job? = null
+
+	private val handler = CoroutineExceptionHandler { _, throwable ->
+		L.w(throwable)
+	}
+
+	private val _savaShareState = MutableStateFlow(false)
+	val savaShareState: StateFlow<Boolean> = _savaShareState
+
+	private val _saveToDefaultList = MutableStateFlow<SaveSharedState<PageTitle>>(Initial())
```

Comment:
What's the reason for having underscore `_` and non-underscore versions of all these fields?

Reply:
The non-underscored fields used only observe values in other entities, such fragments or activities. The underscored field guarantees emit values only from view model. Unfortunately now not other way for it.